File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/templates/TemplateHeroDetails.kt
package com.perrystreet.woof.designsystem.atomic.templates

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.WindowInsetsSides
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.only
import androidx.compose.foundation.layout.safeDrawing
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.layout.SubcomposeLayout
import androidx.compose.ui.unit.Constraints
import com.perrystreet.woof.designsystem.atomic._tokens.sizing.SizingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.PaddingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.SpacingRoles
import com.perrystreet.woof.designsystem.atomic.templates.base.HeroDetailsScrollConnection
import com.perrystreet.woof.designsystem.theme.Theme
import kotlin.math.roundToInt

@Composable
fun TemplateHeroDetails(
    hero: @Composable (dimProgress: () -> Float) -> Unit,
    topBar: @Composable () -> Unit,
    summary: @Composable () -> Unit,
    bottomBar: @Composable () -> Unit,
    overlay: @Composable () -> Unit,
    details: LazyListScope.() -> Unit,
) {
    val topInset = WindowInsets.safeDrawing.only(WindowInsetsSides.Top)
    val horizontalSafe = WindowInsets.safeDrawing.only(WindowInsetsSides.Horizontal)
    val topBarHeight = SizingRoles.InteractionHeight.Comfort.dp + topInset.asPaddingValues().calculateTopPadding()
    val panelHorizontalPadding = PaddingRoles.Screen.Regular.dp
    val panelBottomPadding = PaddingRoles.Screen.Regular.dp
    val bottomBarScrim = Theme.colors.scrimDim
    val listState = rememberLazyListState()
    val connection =
        remember {
            HeroDetailsScrollConnection(
                isListAtTop = { listState.firstVisibleItemIndex == 0 && listState.firstVisibleItemScrollOffset == 0 },
            )
        }
    val progress by remember { derivedStateOf { connection.progress } }

    SubcomposeLayout(
        modifier =
            Modifier
                .fillMaxSize()
                .background(Theme.colors.background),
    ) { constraints ->
        val width = constraints.maxWidth
        val height = constraints.maxHeight
        val loose = Constraints(maxWidth = width, maxHeight = height)

        val bottomBarPlaceables =
            subcompose(HeroDetailsSlot.BottomBar) {
                Box(
                    modifier =
                        Modifier
                            .fillMaxWidth()
                            .background(Brush.verticalGradient(colors = listOf(Color.Transparent, bottomBarScrim)))
                            .imePadding()
                            .navigationBarsPadding()
                            .windowInsetsPadding(horizontalSafe),
                ) {
                    bottomBar()
                }
            }.map { it.measure(loose) }
        val bottomBarHeight = bottomBarPlaceables.maxOfOrNull { it.height } ?: 0
        val cutoutStart = horizontalSafe.getLeft(this, layoutDirection)
        val cutoutEnd = horizontalSafe.getRight(this, layoutDirection)
        val summaryWidth = (width - panelHorizontalPadding.roundToPx() * 2 - cutoutStart - cutoutEnd).coerceAtLeast(0)
        val summaryHeight =
            subcompose(HeroDetailsSlot.Summary, summary)
                .map { it.measure(Constraints(minWidth = summaryWidth, maxWidth = summaryWidth)) }
                .maxOfOrNull { it.height } ?: 0

        val minHeightPx = (summaryHeight + bottomBarHeight + panelBottomPadding.roundToPx()).toFloat()
        val maxHeightPx = (height - topBarHeight.roundToPx()).toFloat().coerceAtLeast(minHeightPx)
        connection.updateBounds(minHeightPx = minHeightPx, maxHeightPx = maxHeightPx)

        val panelHeight = connection.heightPx.coerceIn(minHeightPx, maxHeightPx).roundToInt()
        val bottomBarHeightDp = bottomBarHeight.toDp()

        val panelPlaceables =
            subcompose(HeroDetailsSlot.Panel) {
                LazyColumn(
                    state = listState,
                    modifier =
                        Modifier
                            .windowInsetsPadding(horizontalSafe)
                            .nestedScroll(connection),
                    verticalArrangement = Arrangement.spacedBy(SpacingRoles.Module.Compact.dp),
                    contentPadding =
                        PaddingValues(
                            start = panelHorizontalPadding,
                            end = panelHorizontalPadding,
                            bottom = bottomBarHeightDp + panelBottomPadding,
                        ),
                ) {
                    item(key = SummaryKey) {
                        summary()
                    }
                    item(key = BottomBarSpacerKey) {
                        Spacer(modifier = Modifier.height(bottomBarHeightDp * (1f - connection.progress)))
                    }
                    details()
                }
            }.map { it.measure(Constraints.fixed(width, panelHeight)) }

        val heroPlaceables = subcompose(HeroDetailsSlot.Hero) { hero({ progress }) }.map { it.measure(loose) }
        val topBarPlaceables = subcompose(HeroDetailsSlot.TopBar, topBar).map { it.measure(loose) }
        val overlayPlaceables = subcompose(HeroDetailsSlot.Overlay, overlay).map { it.measure(loose) }

        layout(width, height) {
            heroPlaceables.forEach { it.place(x = 0, y = 0) }
            panelPlaceables.forEach { it.place(x = 0, y = height - panelHeight) }
            topBarPlaceables.forEach { it.place(x = 0, y = 0) }
            bottomBarPlaceables.forEach { it.place(x = 0, y = height - it.height) }
            overlayPlaceables.forEach { it.place(x = 0, y = 0) }
        }
    }
}

private enum class HeroDetailsSlot {
    Hero,
    TopBar,
    Summary,
    Panel,
    BottomBar,
    Overlay,
}

private const val SummaryKey = "hero_details_summary"
private const val BottomBarSpacerKey = "hero_details_bottom_bar_spacer"

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Templates/TemplateHeroDetails.swift
import SwiftUI

public struct TemplateHeroDetails<Hero: View, TopBar: View, Summary: View, BottomBar: View, Overlay: View, Details: View>: View {
    @Environment(\.theme) private var theme
    @Environment(\.templateSafeAreaInsets) private var inheritedInsets
    @State private var scrollOffset: CGFloat = 0
    @State private var summaryHeight: CGFloat = 0
    @State private var bottomBarHeight: CGFloat = 0

    private let hero: (CGFloat) -> Hero
    private let topBar: () -> TopBar
    private let summary: () -> Summary
    private let bottomBar: () -> BottomBar
    private let overlay: () -> Overlay
    private let details: () -> Details

    public init(
        @ViewBuilder hero: @escaping (CGFloat) -> Hero,
        @ViewBuilder topBar: @escaping () -> TopBar,
        @ViewBuilder summary: @escaping () -> Summary,
        @ViewBuilder bottomBar: @escaping () -> BottomBar,
        @ViewBuilder overlay: @escaping () -> Overlay,
        @ViewBuilder details: @escaping () -> Details
    ) {
        self.hero = hero
        self.topBar = topBar
        self.summary = summary
        self.bottomBar = bottomBar
        self.overlay = overlay
        self.details = details
    }

    public var body: some View {
        GeometryReader { proxy in
            let topInset = max(proxy.safeAreaInsets.top, inheritedInsets.top)
            let bottomInset = max(proxy.safeAreaInsets.bottom, inheritedInsets.bottom)
            let containerHeight = proxy.size.height + proxy.safeAreaInsets.top + proxy.safeAreaInsets.bottom
            let topBarHeight = SizingRoles.InteractionHeight.comfort.rawValue + topInset
            let panelBottomPadding = PaddingRoles.Screen.regular.rawValue
            let panelMinHeight = summaryHeight + bottomBarHeight + panelBottomPadding
            let maxOffset = max(containerHeight - topBarHeight - panelMinHeight, 1)
            let progress = min(max(scrollOffset / maxOffset, 0), 1)
            ZStack(alignment: .top) {
                hero(progress)
                    .frame(width: proxy.size.width, height: containerHeight)
                ScrollView {
                    VStack(alignment: .leading, spacing: SpacingRoles.Module.compact.rawValue) {
                        Color.clear
                            .frame(height: maxOffset + topBarHeight)
                        summary()
                            .onGeometryChange(for: CGFloat.self) { $0.size.height } action: { summaryHeight = $0 }
                        Color.clear
                            .frame(height: bottomBarHeight * (1 - progress))
                        details()
                    }
                    .padding(.horizontal, PaddingRoles.Screen.regular.rawValue)
                    .padding(.bottom, bottomBarHeight + panelBottomPadding)
                }
                .scrollIndicators(.hidden)
                .scrollTargetBehavior(HeroDetailsScrollConnection(maxOffset: maxOffset))
                .onScrollGeometryChange(for: CGFloat.self) { $0.contentOffset.y + $0.contentInsets.top } action: { _, offset in
                    scrollOffset = offset
                }
                .frame(width: proxy.size.width, height: containerHeight)
                topBar()
                    .padding(.top, topInset)
                VStack(spacing: 0) {
                    Spacer(minLength: 0)
                    bottomBar()
                        .padding(.bottom, bottomInset)
                        .onGeometryChange(for: CGFloat.self) { $0.size.height } action: { bottomBarHeight = $0 }
                        .background {
                            LinearGradient(
                                colors: [Color.clear, theme.colors.scrimDim],
                                startPoint: .top,
                                endPoint: .bottom
                            )
                        }
                }
                .frame(height: containerHeight)
                overlay()
            }
            .frame(width: proxy.size.width, height: containerHeight)
            .ignoresSafeArea(.container)
        }
        .background(theme.colors.background.ignoresSafeArea())
    }
}
